### PR TITLE
A first implementation of userdata in callbacks.

### PR DESCRIPTION
**Pull request by dddomodossola**
A first implementation of userdata in callbacks. Just an example in set_on_click_listener, see its usage in simple_app.py

### CODE REVIEW

**Comment by dddomodossola** [comment]
@nzjrs Give a look at simple_app.py. I implemented the usedata only for set_on_click_listener, just for an example. Is this ok?

File: remi/gui.py [hunk 1]
```diff
@@ -537,7 +538,7 @@ def set_on_mousedown_listener(self, listener, funcname):
     def onmouseup(self, x, y):
         """Called when the user releases left or right mouse button over a Widget.
         """
-        return self.eventManager.propagate(self.EVENT_ONMOUSEUP, [x, y])
+        return self.eventManager.propagate(self.EVENT_ONMOUSEUP, (x, y))
```
Comment (nzjrs) [inline remi/gui.py]: In this case, x and y are not userdata, but are the defined by the callback API.

Comment (dddomodossola) [inline remi/gui.py]: Yeah, I've not passed yet userdata in other callbacks, it is just implemented for set_on_click_listener. The change you highlighted is just a conversion from list to tuple and is not directly related to the userdata. Is this clear?

**Comment by dddomodossola** [comment]
Ok for the widget retrieved in the callback, I will do this. But It's not clear for me the problem of mixing userdata and standard parameter. You showed two code examples:

```python
def my_clicked_callback(self, widget_that_was_clicked, *userdata):
    pass

def my_onmousemove_callback(self, widget, x, y, *userdata):
    pass
```

Are these the desired callbacks prototypes? In this case *userdata is not unpacked. For me could be ok.

**Comment by nzjrs** [comment]
Yeah, it's clear, it just reminded me to mention this....

**Comment by nzjrs** [comment]
The usual convention is that userdata is always unpacked when calling the
callback, and if the user passes userdata parameters, he is responsible for
making his prototype API compatible.

IMO the API is nicer and It is easier to work with if the eventmanager
unpacks userdata - but it can make the implementation more complicated.

**Comment by dddomodossola** [comment]
The changes to pass unpacked userdata to callbacks are already implemented in EventManager.
i.e.:
`self.bt.set_on_click_listener(self, 'on_button_pressed', 'userdata1', 2)`

and:

```python
def on_button_pressed(self, userdata1, userdata2):
    pass
```

As you can see the 'userdata1' and '2' are additional userdata parameters to the listener registration. The listener receives userdata1 and userdata2 unpacked.

This should be conformant. Isn't  it?

The missing thing is the widget emitter parameter that should be passed to the listener. So the result of this should be:
`self.bt.set_on_click_listener(self, 'on_button_pressed', 'userdata1', 2)`

and:

```python
def on_button_pressed(self, widget, userdata1, userdata2):
    pass
```

Is this right? Or I misunderstood something?

**Comment by nzjrs** [comment]
I think a more interesting test is to see how the API works for a callback
that passes it''s own parameters in addition. Maybe it is clearer what I
mean if you try to implement the onmouseup / onmousedown callbacks, with
optional userdata, in the simple_app demo. I hope that clarifies things a
bit.

**Comment by dddomodossola** [comment]
@nzjrs Ok I test it ;-)

**Comment by dddomodossola** [comment]
Take a look at the last commit. This is an example with onmousedown and onmouseup. Seems ok to me. What about this?

**Comment by nzjrs** [comment]
How do you feel about breaking API and pass the widget that sent the event
as the first parameter to all callbacks?

**Comment by dddomodossola** [comment]
I feel ok with this. It will be usefull in different use cases. Can I proceed on implementing this additional change?

**Comment by nzjrs** [comment]
I added another example. Can you quickly prototype passing the sender as
the first argument?

If we are breaking API, how about instead of passing the function name, we
pass the actual function, or is there a reason this is not currently done?

**Comment by dddomodossola** [comment]
Yeah, I will try to make it soon.

You mean passing the "function pointer"? This causes serious difficulties in the Editor application, but maybe I can solve this. Let me check this evening about this.

**Comment by nzjrs** [comment]
Yeah. Passing the function by name is quite weird when compared to other
gui toolkits

**Comment by dddomodossola** [comment]
This is true. I will do the best to implement this. It allows also the error checking and autocompletion by IDEs.

**Comment by dddomodossola** [comment]
@nzjrs Is the last commit conformant to what you suggested?

**Comment by nzjrs** [comment]
I changed the example a bit to be idiomatic of the benefits of passing
userdata and the sender as the first argument.


I had forgotten that one can only have one callback per signal. Do you
think it would be possible to also fix this limitation too?

then, all the functions get renamed from set_XXX_listener ->
add_XXX_listener

**Comment by dddomodossola** [comment]
I think that this is not feasible John. I feel it logically wrong and I think this is not allowed by the architecture of the library. ,Consider that some callbacks could return a value to the page, acting as ajax does...

**Comment by dddomodossola** [comment]
John, what about new changes? :-D

**Comment by nzjrs** [comment]
> I think that this is not feasible John. I feel it's logically wrong and I think this is not allowed by the 
> architecture of the library. ,Consider that some callbacks could return a value to the page, acting as 
> ajax does...

As you wish. It's a common pattern to have multiple callbacks in other GUI frameworks, but if it's not possible then I will just do it in the code on my side.

The rest of the changes look good. I would say you could squash the branch and merge it.

**Pull request merged**